[Devops-890] nixify plutus

disassembler opened the pull request:
Building plutus with nix for automated hydra CI testing.

cleverca22 commented on default.nix:
```diff
@@ -0,0 +1,61 @@
+      (type == "symlink" && lib.hasPrefix "result" baseName) ||
+      (type == "directory" && baseName == ".stack-work")
+    );
+  source = builtins.filterSource cleanSourceFilter ./.;
```
this filter is unused

disassembler replied:
ah, I stopped using it as a troubleshooting step with the hydra issue we were debugging, back in.

cleverca22 commented on fetch-nixpkgs.nix:
```diff
@@ -0,0 +1,8 @@
+let
+  # temporary hack until scripts/nix-shell.sh ceases to use -p
```
we could just skip using this hack and never use `-p` to begin with in this repo

disassembler replied:
I simplified it significantly. It now assumes nix 2 is available and fetch-nixpkgs and fetchNixpkgs are a single file again.

cleverca22 replied:
the whole `runCommand` and symlink part could also be removed, since thats only needed to allow you to `nix-build` the derivation and insert into `NIX_PATH`

cleverca22 commented on lib.nix:
```diff
@@ -0,0 +1,19 @@
+let
+  # Allow overriding pinned nixpkgs for debugging purposes via cardano_pkgs
+  fetchNixPkgs = let try = builtins.tryEval <cardano_pkgs>;
```
this still says cardano

cleverca22 commented on shell.nix:
```diff
@@ -0,0 +1,27 @@
+
+let 
+  plutusPkgs = import ./. {};
+  ghc = pkgs.haskell.packages.ghc822.ghcWithPackages (ps: [
```
this is using the unpatched ghc, rather then the patched ghc

disassembler replied:
yeah... do we need the patched here? We're not building binaries for OSX currently either, we're just running tests on linux. Maybe I just pull that attribute out completely and we revisit later if we do need to patch GHC for OSX building. That's what I did in my most recent commit.

disassembler commented:
Thanks for the review @mchakravarty ! I think we should move away from multiple stack.yaml files as soon as we can. It makes it much harder to manage the CI scripts as new things are added. The way the pkgs/generate.sh script works is it reads a stack.yaml file (which can have multiple packages, see the updated stack.yaml in the root, or better yet the one used by cardano-sl) and it builds dependency trees for all the packages, and then generates a nix equivalent. This means whenever the cabal or stack.yaml file is updated (say to add an additional package), it will  create a nix equivalent when pkgs/generate.sh is ran so that can be used by anyone using nix or the hydra CI system.

My intention is to run CI on hydra, not buildkite, as there are no build impurities here. Hydra is like buildkite, except it runs the build in a completely isolated sandbox using nix. You can see the two updated build statuses above linking to the hydra builds. It also caches the build to s3 so nix users don't have to build it locally, they can just pull the pre-built version from the binary cache.

If we really don't want the prototype building in hydra, I suggest I just remove it from release.nix but leave the nix files there in case anyone wants to use the prototype.

As for OSX, I should clarify here, only the hedgehog and shellcheck tests that are reporting build statuses back to github are running on linux. The build itself is building binaries for osx and linux: https://hydra.iohk.io/eval/245195

disassembler commented:
I moved the plutus-prototype files to it's own subdirectory, and updated the stack.yaml to point there instead '.'. I also removed the stack.yaml from the `language-plutus-core` subdirectory. For the new package being added later this week, they should do similar, add the project to the root `stack.yaml` and keep the cabal and source files in a subdirectory.  This should illustrate how we should be handling multiple projects in a single git repository.

mchakravarty commented:
@disassembler Thank you very much for the explanation! It makes perfect sense, and I agree that this is the right way forward.
